### DISCUSSION TITLE
How do I detect clicks on the background of a layout with buttons and labels inside?

### DISCUSSION THREAD

**Discussion by CanisLupus** [emilk/egui#4856] [Q&A] [answered]
Hi! I'm struggling to create an area with buttons/labels inside, whose background I can also click (to do something custom).

I know about `Sense` but it can't be added to a layout, a frame, or a scope. I've tried an Area with disabled drag, but it doesn't expand to the size of its contents. I've tried detecting clicks on rectangular area myself but it will catch all the clicks, including to buttons in the area.

Do you know how I can achieve this? Thanks!

**Comment by CanisLupus** [comment, marked as answer]
I accidentally stumbled upon `interact_bg` just now 😅

You can call it with a `Sense` inside the widget you are drawing, and the background will detect that `Sense`:

```
ui.vertical(|ui| {
    // draw some buttons, labels, etc

    if ui.interact_bg(egui::Sense::click()).clicked() {
        // click-related code
    }
});
```

Keep in mind that `interact_bg` works with the ID of the parent element you are drawing, so if you have multiple ones they will each require a custom ID, which you can give them with `push_id`.

**Comment by hydra** [comment]
apparently this is now deprecated, an example showing how to detect clicks on the background of a frame using the non-deprecated alternative would be nice.  Here's the PR that deprecates it: https://github.com/emilk/egui/pull/5054

**Reply by lucasmerlin** [reply]
There is an example called interactive_container that showcases it

**Reply by hydra** [reply]
yes, I found that in PR #5054 and based my approach on it.  thanks!

**Reply by hydra** [reply]
I also note that the deprecation message on `interact_bg` was not helpful, since it doesn't mention `ui.scope_builder`.

To find out what to do I had to trawl the codebase looking for usages of `Ui::sense` and stumbled on the code in the PR.

I'm also not sure what 'Ui::response' is referring to, since I couldn't find a method on `Ui` called `response`.

Perhaps you could do another PR to clarify the deprecation message?

**Reply by lucasmerlin** [reply]
There definietly is a `Ui::response` 🤔: https://docs.rs/egui/latest/egui/struct.Ui.html#method.response

**Comment by hydra** [comment]
just following up, this example shows how to detect clicks on en egui::Sides.

```rust
let response = ui.scope_builder(
    UiBuilder::new()
        .sense(Sense::click()),
    |ui| {
        egui::Frame::NONE
            .inner_margin(Margin::symmetric(4, 0))
            .show(ui, |ui| {
                egui::Sides::new()
                    .height(ui.available_height())
                    .show(
                        ui,
                        |ui| {
                            Label::new(RichText::new("LEFT")
                                .monospace().color(Color32::RED)).selectable(false).ui(ui);
                        },
                        |ui| {
                            Label::new(RichText::new("RIGHT")
                                .monospace().color(Color32::GREEN)).selectable(false).ui(ui);
                        }
                    );
            });
    }).response;

if response.clicked() {
    debug!("clicked");
}
```

hope this helps someone, it took me about 30 attempts yesterday to make this work. I found accomplishing this extremely non-intuitive.

**Reply by TigranKirakosov** [reply]
@hydra thanks a lot! I was on a verge of loosing my mind while attempting to achieve this thing.

**Reply by Sheldonfrith** [reply]
Thanks as well! I agree its currently a very unintuitive api. This is what I had to do to get focus-on-click for a simple horizontal section with a dropdown menu... 

```rust
let horiz_response = ui.scope_builder( UiBuilder::new().sense(Sense::click()),|ui| {
            let horiz_response = ui.horizontal(|ui| {
            let label_response = ui.label(label);

            let combo_box_response = egui::ComboBox::from_label("Select an option")
                .selected_text(value.clone())
                .show_ui(ui, |ui| {
                    if let Some(options) = &form_state.possible_values {
                        for option in options {
                            if ui
                                .selectable_value(&mut new_value, option.clone(), option)
                                .clicked()
                            {
                                new_value = option.clone();
                                changed = true;
                            }
                        }
                    }
                });
            combo_box_response.response.clicked() || label_response.clicked()
        });
        horiz_response.response.clicked() || horiz_response.inner
    });
        if horiz_response.response.clicked() || horiz_response.inner{
            println!("clicked on dropdown");
            ui.memory_mut(|m| m.request_focus(horiz_response.response.id));
        }
```
Pretty jarring compared to my experience with non-rust UI libraries.
